arch/arm/arm64: Introduce Pointer Authentication support

michpappas opened the pull request:
### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://docs.unikraft.org/contribute.html) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Base target

 - Architecture(s): arm64
 - Platform(s): common
 - Application(s): N/A

### Additional configuration

This PR introduces `CONFIG_ARM64_FEAT_PAUTH` to enable Pointer Authentication support.

### Description of changes

Pointer Authentication (PAuth) allows signing and authenticating pointers to harden the system against classes of attacks that rely on the manipulation of pointers, such as ROP. Pointer Authentication Codes (PACs) are created by signing a pointer and a 64-bit modifier with an 128-bit key. The modifier is a value that is normally used to restrict the PAC to a specific context (eg the SP). Keys are stored in registerers.

The PAC is stored in the unused upper bits of the memory address, and can be later verified to ensure that the pointer has not been tampered.

The reference algorithm is [the QARMA block cipher](https://eprint.iacr.org/2016/444.pdf), but it is possible for architectures to use an IMPLEMENTATION DEFINED algorithm instead.
 
**Basic Instructions:**
* PAC*: Sign address
* AUT*: Authenticate address
* XPA*: Strip PAC from address (useful for things like unwinding)

**Combined Instructions:**
* RETA*:  Auth and return
* BRA*:   Auth and branch
* BLRA*:  Auth and branch-link
* ERETA*: Auth and return from exception
* LDRA*:  Auth and load

**Keys:**
* APIA_Key: Key A for signing instructions
* APIB_Key: Key B for signing instructions
* APDA_Key: Key A for signing data
* APDB_Key: Key B for signing data
* APGA_Key: Key A for generic use

For more background see:
* [The Arm Architecture Reference Manual Armv8, for Armv8-A architecture profile](https://developer.arm.com/documentation/ddi0487/ga)
* [Learn the architecture: Providing protection for complex software](https://developer.arm.com/documentation/102433/latest/)
* [Pointer Authentication on ARMv8.3](https://www.qualcomm.com/media/documents/files/whitepaper-pointer-authentication-on-armv8-3.pdf)

## Architecture support

**Armv8.3:**
* Introduce FEAT_PAuth as a mandatory feature.
* Introduce FEAT_PAuth2 as an optional feature in Armv8.3 and mandatory in
  Armv8.6. FEAT_PAuth2 changes the way the PAC is applied to the address
  (XOR vs replace).
* Introduce FEAT_FPAC, which allows to control the faulting behavior when authentication
  fails. FEAT_FPAC is an optional feature and depends on FEAT_PAuth2.

## GCC support

**GCC 7:**
* Introduce support for armv8.3-a
* Introduce `--msign-return-address=[none|non-leaf|all]`

The parameters passed to `--msign-return-address` are interpreted as:
* `none`: Do not sign return addresses
* `non-leaf`: Sign/auth the return address of non-leaf functions.
* `all`: Sign/auth the return address of non-leaf and leaf functions.

GCC implements the Basic Set of PAuth instructions using the HINT instruction, in the so-called NOP space. This allows executing protected binaries on older platforms that do not implement Armv8.3-a. In these platforms PAuth instructions
execute as NOP.

When `sign-return-address` is enabled without setting `-march=armv8.3-a`, GCC generates PACIASP / AUTIASP sequences that sign and authenticate the LR upon function entry and exit. In the following snippet notice the opcode of the
HINT instruction (0xd5032300) that GCC generates for PACIASP and AUTIASP.
```
0000000000400564 <main>:
  400564:       d503233f        paciasp
  400568:       a9bf7bfd        stp     x29, x30, [sp, #-16]!
  40056c:       910003fd        mov     x29, sp
  400570:       90000000        adrp    x0, 400000 <_init-0x3e8>
  400574:       9118c000        add     x0, x0, #0x630
  400578:       97ffffb6        bl      400450 <puts@plt>
  40057c:       52800000        mov     w0, #0x0
  400580:       a8c17bfd        ldp     x29, x30, [sp], #16
  400584:       d50323bf        autiasp
  400588:       d65f03c0        ret
```

When `-msign-return-address` is used along with `-march=armv8.3-a`, GCC
generates PACIASP / RETAA sequences instead. In the snippet below notice the
opcode of RETAA (0xd65f0bff) which is in the fused space (instructions in the
Combined Set do not have HINT implementations). Clearly, this code cannot be
executed in architectures earlier than Armv8.3-a.
```
0000000000400564 <main>:
  400564:       d503233f        paciasp
  400568:       a9bf7bfd        stp     x29, x30, [sp, #-16]!
  40056c:       910003fd        mov     x29, sp
  400570:       90000000        adrp    x0, 400000 <_init-0x3e8>
  400574:       9118a000        add     x0, x0, #0x628
  400578:       97ffffb6        bl      400450 <puts@plt>
  40057c:       52800000        mov     w0, #0x0
  400580:       a8c17bfd        ldp     x29, x30, [sp], #16
  400584:       d65f0bff        retaa
```

**GCC 9:**
* Introduce support for armv8.5-a
* Deprecate `-msign-return-address`
* Introduce `-mbranch-protection=[none|pac-ret{+leaf}|bti|standard]`
* Introduce `--enable-standard-branch-protection` as a short to `mbranch-protection=standard`

The parameters passed to `-mbranch-protection` are interpreted as:
* `none`: Disables all protections
* `pac-ret`: Enables PAuth for function returns on non-leaf functions. The
           `+leaf` modifier enables protection for leaf functions.
* `bti`: Enables Branch Target Identification
* `standard`: Enables all protections

**GCC 10:**
* Introduce the b-key addendum to pac-ret in `--mbranch-protection`.
  This allows using the APIB_Key instead of the APIA_Key when signing
  return pointers.

## Changes introduced in this series

This PR provides platforms with an API to initialize PAuth keys, and enable PAuth.

A platform can initialize PAuth as:
```
#ifdef CONFIG_ARM64_FEAT_PAUTH
	if (ukplat_pauth_enable())
		UK_CRASH("Pointer Authentication is not available");
#endif
```

To generate PAyth keys, it is required that platforms provide an implementation of the key generation function that uses an adequate source of randomness.
```
void ukplat_pauth_gen_key(__u64 *key_hi, __u64 *key_lo);
```

Platforms that support Armv8.5 can use the random number generation instructions introduced into the architecture (`RNDR`/`RNDRSS`). Others can initialize drivers of an HWRNG/TRNG, or request randomness from the TEE.

Since the above functions are used for the initialization of PAuth, it is required that GCC excludes them when generating PACIASP/RETAA sequences. Otherwise, once PAuth is enabled, authentication will fail upon return. A macro is provided to help overriding the global gcc settings per function:
```
#define __no_pauth __attribute__((target("branch-protection=none")))
```

This macro is used by `pauth_enable()`, and should be used by all functions generated by GCC in the boot chain, up to - and including - the caller of `pauth_init()`.

Additionally to the above, the following choices are made:

**Do not enable backwards compatibility mode.**
If `-march=armv8.3-a` is not set, it is not possible to initialize the keys.
GCC exits with the following error:
```
bash$ ~/toolchains/gcc-arm-10.3-2021.07-x86_64-aarch64-none-elf/bin/aarch64-none-elf-gcc -mbranch-protection=standard /tmp/test.c
/tmp/ccx7K1nD.s: Assembler messages:
/tmp/ccx7K1nD.s:17: Error: selected processor does not support system register name 'apiakeyhi_el1'
```
I believe that Arm's intention behind compatibility mode was mostly to provide backwards compatibility on userspace libraries. Clearly on Unikraft this feature cannot be utilized, except from the case of linking with prebuilt libraries.
Because of that this PR sets `-march=armv8.3-a` when enabling `CONFIG_ARM64_FEAT_PAUTH`.

**Enable leaf protection by default**

Although it's not strictly required as leaf functions do not save the LR, it doesn't hurt to be there as an additional protection.

**Initialize only the PACIA_Key**

The motivation for that is minimizing the impact on boot time. Additional keys can be enabled later, if needed.

**Do not introduce additional APIs for sign / auth / strip**

The intention of this PR is to provide initial PAuth support. As such, its scope is limited to the signing and verification of the function return address. Nevertheless, Pointer Authentication provides the possibility to harden the system further by signing more pointers. Additional functionality can be introduced into the Unkraft API later, on demand.

michpappas commented:
This PR depends on #368

craciunoiuc commented:
Hey @michpappas,

I did an initial pass of everything and the PR seems good!

I will come back to it to test it once the other PR is merged.

Cezar

michpappas commented:
Rebased and fixed a bug in the `__no_branch_prediction` macro in `plat/common/include/arm/arm64/pauth.h`.

craciunoiuc commented on plat/common/include/arm/arm64/pauth.h:
```diff
@@ -0,0 +1,72 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/* Macro to disable the global branch-protection settings for a specific
+ * function. This needs to be declared on any functions in the boot-chain
+ * up to the caller of ukplat_pauth_init() to avoid PAuth errors upon return.
+ */
+#if CONFIG_ARM64_FEAT_PAUTH
+#define __no_branch_protection __attribute__((target("branch-protection=none")))
+#else
+#define __no_branch_protection
+#endif
```
I don't really get the `#if` brace here.

1. There might be other cases where branch protection needs to be disabled apart from the `PAUTH` case

2. This macro is in the `pauth.h` file. It should be used only in the context of `pauth` anyway.

michpappas replied:
Hi @craciunoiuc, 

> 1. There might be other cases where branch protection needs to be disabled apart from the PAUTH case

Yes in general the `branch-protection` option can also be used for BTI, so my on-off logic was in the current scope of BTI not supported yet. But that should change soon, as I have a PR on BTI almost ready.
With BTI enabled the macro will have to include additional logic to take into account whether `CONFIG_ARM64_FEAT_BTI` is enabled or not, and change the value of `branch-protection` accordingly. Saying that, a more suitable name for that macro would be something like `_no_pauth_prot` or something similar.

> 2. This macro is in the pauth.h file. It should be used only in the context of pauth anyway.

Well, it depends, one could include the header unconditionally,  but overall I agree. Especially as GCC uses the same parameter for PAuth and BTI, this should probably be moved to a generic header.

I'll think about it a bit more as I work on the BTI PR. Will push an update soon.

craciunoiuc reviewed (commented):
Hey @michpappas,

This PR looks good. I have just a small comment. After that I can approve it 🙂

michpappas committed:
arch/*: Introduce uk/asm/compiler.h

Introduce a header for architecture-specific compiler definitions.
Architectures add compiler definitions in uk/asm/compiler.h, which
are made available through uk/essentials.h

This commit adds empty headers for arm, arm64, and x86_64.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas committed:
arch/arm/arm64: plat/common/arm: Add Pointer Authentication support

Pointer Authentication (FEAT_PAuth) allows signing and authenticating
pointers to harden systems against classes of attacks that rely on
memory corruption, such as stack smashing and ROP.

Pointer Authentication Codes (PACs) are created by signing a pointer and
a 64-bit modifier with an 128-bit key. The modifier is a value that is
used to restrict the PAC to a specific context (eg the SP). Keys are
stored in registers. The PAC is stored in the unused upper bits of the
memory address, and can be later verified to ensure that the pointer has
not been tampered with.

This commit adds basic Pointer Authentication support for platforms that
support arm64. Specifically, it introduces:

- The CONFIG_ARM64_FEAT_PAUTH option. When enabled, GCC is instructed
  to generate PACIASP/RETAA sequences to protect the return address of
  functions using the PACIA_Key.

- An API to set up and enable Pointer Authentication. Platforms are
  required to implement a key generation function that uses a secure
  source of randomness.

- A macro to disable Pointer Authentication in certain functions. This
  is required on any returning function that is called prior to the
  initialization of Pointer Authentication.

Unikraft images generated with FEAT_PAuth are compatible with
architectures that implement Armv8.3 and above. That is due to the fact
that, although the basic set of Pointer Authentication instructions is
implemented in the NOP space, combined instructions and registers
holding the keys are not.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas commented:
## Changes in this update

**Config.uk:**
* Switch to a flat structure
* Update help message to a more generic description

**Makefile.uk:**
* Add comment about minimum GCC version
* Rework in anticipation of additional extensions (see #421)

**compiler.h:**
* Introduce header for architecture-specific compiler definitions.
* Move the `__no_branch_protection` macro to compiler.h.
* Rename the `__no_branch_protection` macro to `__no_pauth`.

**pauth.c:**
* Reduce API to a single function, `ukpauth_init()`

**pauth.h:**
* Add include guards
* Change the return type of `ukplat_gen_key()` to void
* Add note about using `__no_pauth` when declaring `ukplat_gen_key()`

**lcpu.h:**
* Use the _AC macro where needed

michpappas commented:
PR description and commit message updated based on latest changes.

craciunoiuc commented on arch/arm/arm64/Config.uk:
```diff
@@ -145,6 +145,18 @@ choice
+	help
+		Enable signing and authentication of pointers. This
+		provides protections against classes of attacks that
+		rely on memory corruption, such as smash stacking and
```
```suggestion
		rely on memory corruption, such as stack smashing and
```

craciunoiuc reviewed (changes requested):
Hello again @michpappas,

The PR looks good now, but before adding my tag, here are some typos from the second commit message 😅:

"stored in `registerers`. The PAC" -> "stored in registers. The PAC"
"been tampered." -> "been tampered with."
"Pointer `Authenticaiton`" ->"Pointer Authentication"
"`equired` to" -> "required to"

After this, we can finally close this PR 😆.

craciunoiuc commented on arch/arm/arm64/Config.uk:
```diff
@@ -145,6 +145,18 @@ choice
 			Compile and optimize for Neoverse V1 CPUs.
 endchoice
 
+menu "Armv8-A Extensions"
+	depends on MCPU_ARM64_NONE
```
Another thing:

Why does this feature depend on `MCPU_ARM64_NONE`?
As far as I understand, doesn't this mean "no specific processor" or something like that?

michpappas replied:
Hi @craciunoiuc, that is because if you choose a specific core, GCC enforces the architecture revision (eg Armv8-3) of that IP. On the other hand, Arm has not yet released IP that implements newer architectural revisions. For example, you won't find a Cortex-A that supports Armv8.3. There are however implementations of licensees, such as Apple, yet these cores are not known to GCC. Consequently, to be able to free-select your own configuration, either for an existing core that is not known to GCC, or an emulated core such as QEMU's virt, you need to use `-mcpu=none`. The changes in 1fadd78ae2b46f03c0b3eb807311ccd6ca4dd743 were made to address the very same issue.

craciunoiuc replied:
Understood 👍

michpappas commented:
@craciunoiuc, typos are fixed now

craciunoiuc reviewed (approved):
All good!

@razvand your turn now 😉 

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@gmail.com>

razvand commented on arch/arm/arm/include/uk/asm/compiler.h:
```diff
@@ -0,0 +1,35 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) YYYY, Copyright Holder. All rights reserved.
```
@michpappas , please add the date and your name here.

craciunoiuc replied:
@razvand same as the other comment

razvand commented on arch/x86/x86_64/include/uk/asm/compiler.h:
```diff
@@ -0,0 +1,35 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) YYYY, Copyright Holder. All rights reserved.
```
Proper date + name

craciunoiuc replied:
I think this is like this because the files are empty and the person that will fill in code will write themselves there.

michpappas replied:
@razvand, what Cezar said. These are empty, so I'm not sure if it makes any sense to copyright to myself.

razvand reviewed (commented):
@michpappas , this looks good. Before approving, please check the copyright information.

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvan.deaconescu@cs.pub.ro>
